test/end2end_test.go

```diff
@@ -1595,6 +1595,61 @@ func testExceedMaxStreamsLimit(t *testing.T, e env) {
 	}
 }
 
+func TestStreamsQuotaRecovery(t *testing.T) {
+	defer leakCheck(t)()
+	for _, e := range listTestEnv() {
+		testStreamsQuotaRecovery(t, e)
+	}
+}
+
+func testStreamsQuotaRecovery(t *testing.T, e env) {
+	te := newTest(t, e)
```

Comment:
This is almost the only way to test the tricky race condition and something similar like this because there is no deterministic way to reproduce that. Typically, I run the tests a thousand times to detect any flakeness before submitting.